bftengine/src/bftengine/ClientsManager.cpp

```diff
@@ -88,33 +90,41 @@ uint32_t ClientsManager::numberOfRequiredReservedPages() const { return required
     const uint32_t firstPageId = clientIdx * reservedPagesPerClient_;
 
-    if (!stateTransfer_->loadReservedPage(resPageOffset() + firstPageId, sizeOfReservedPage_, scratchPage_)) continue;
-
-    ClientReplyMsgHeader* replyHeader = (ClientReplyMsgHeader*)scratchPage_;
-    ConcordAssert(replyHeader->msgType == 0 || replyHeader->msgType == MsgCode::ClientReply);
-    ConcordAssert(replyHeader->currentPrimaryId == 0);
-    ConcordAssert(replyHeader->replyLength >= 0);
-    ConcordAssert(replyHeader->replyLength + sizeof(ClientReplyMsgHeader) <= maxReplySize_);
-
-    // YS TBD: Multiple replies for client batching should be sorted by incoming time
-    auto& repliesInfo = indexToClientInfo_.at(clientIdx).repliesInfo;
-    if (repliesInfo.size() >= maxNumOfReqsPerClient_) deleteOldestReply(clientId);
-    const auto& res = repliesInfo.insert_or_assign(replyHeader->reqSeqNum, MinTime);
-    const bool added = res.second;
-    LOG_INFO(CL_MNGR, "Added/updated reply message" << KVLOG(clientId, replyHeader->reqSeqNum, added));
-
+    // load all replies into the map of replies one by one
+    for (int i = 0; i < maxNumOfReqsPerClient_; i++) {
+      if (!stateTransfer_->loadReservedPage(
+              resPageOffset() + firstPageId + i, sizeOfReservedPage_, scratchPage_.data() + i * sizeOfReservedPage_))
+        continue;
+
+      auto replyHeader = (ClientReplyMsgHeader*)(scratchPage_.data() + i * sizeOfReservedPage_);
+      auto replyPtr = std::make_shared<ClientReplyMsg>(myId_, replyHeader->replyLength);
+      ConcordAssert(replyHeader->msgType == 0 || replyHeader->msgType == MsgCode::ClientReply);
+      ConcordAssert(replyHeader->currentPrimaryId == 0);
+      ConcordAssert(replyHeader->replyLength >= 0);
+      ConcordAssert(replyHeader->replyLength + sizeof(ClientReplyMsgHeader) <= maxReplySize_);
+
+      auto& repliesInfo = indexToClientInfo_.at(clientIdx).repliesInfo;
+      if (repliesInfo.size() >= maxNumOfReqsPerClient_) deleteOldestReply(clientId);
+      const auto& res = repliesInfo.insert({replyHeader->reqSeqNum, replyPtr});
+      const bool added = res.second;
+      LOG_INFO(CL_MNGR, "Added/updated reply message" << KVLOG(clientId, replyHeader->reqSeqNum, added));
+    }
+    // get the newest reply from the map of replies.
+    auto lastReply = indexToClientInfo_.at(clientIdx).repliesInfo.rbegin();
+    auto lastReplyHeader = lastReply->second->b();
     // Remove old pending requests
     auto& requestsInfo = indexToClientInfo_.at(clientIdx).requestsInfo;
     for (auto it = requestsInfo.begin(); it != requestsInfo.end();) {
-      if (it->first <= replyHeader->reqSeqNum) {
-        LOG_INFO(CL_MNGR, "Remove old pending request" << KVLOG(clientId, replyHeader->reqSeqNum));
+      if (it->first <= lastReplyHeader->reqSeqNum) {
+        LOG_INFO(CL_MNGR, "Remove old pending request" << KVLOG(clientId, lastReplyHeader->reqSeqNum));
         it = requestsInfo.erase(it);
       } else
         it++;
```

Comment:
I believe we can just break here, since the map is sorted

Reply:
Done, thanks.